File: node/src/agent.ts
/**
 * `HawcxAgent` — the customer-facing entry point.
 *
 * Thin ergonomic wrapper around `AssemblerClient`. Takes a socket path to an
 * already-running Assembler agent endpoint, performs the IPC handshake, and
 * exposes `invoke` for sending tool calls.
 *
 * Per CS v6.7.4 §39, the SDK process never sees session keys — the Assembler
 * handles all crypto. The SDK only carries plaintext request bodies inbound
 * and decrypted response bodies outbound, both over the local IPC channel.
 *
 * See the package-level README for the prerequisites (supervisor running,
 * agent identity provisioned via CAA).
 */

import { Buffer } from "node:buffer";
import { randomUUID } from "node:crypto";
import * as os from "node:os";
import * as path from "node:path";
import * as process from "node:process";

import {
  AssemblerClient,
  TokenTransport,
  type ToolCallResponse,
} from "./ipc";

export interface HawcxAgentInvokeOptions {
  targetRsUrl: string;
  httpMethod?: string;
  headers?: Record<string, string>;
  tool?: string;
  action?: readonly string[];
  resource?: string;
  constraints?: Record<string, unknown>;
  body?: Buffer | null;
  claimedIntentHash?: string;
  toolArguments?: unknown;
  contentType?: string;
  transport?: TokenTransport;
  requestId?: string;
}

function defaultIpcDir(): string {
  if (process.platform === "win32") {
    // Windows pipes are in the kernel namespace; ipc_dir is unused.
    return "";
  }
  const xdgRuntime = process.env.XDG_RUNTIME_DIR;
  if (xdgRuntime) return path.join(xdgRuntime, "hawcx");
  return path.join(os.tmpdir(), "hawcx");
}

/**
 * Compute the conventional Assembler agent-socket path for an agent id.
 *
 * - **Unix:** `{ipc_dir}/{agentId}/agent-assembler-{index}.sock`
 * - **Windows:** `\\\\.\\pipe\\haap-{agentId}-agent-assembler-{index}`
 */
export function defaultEndpointFor(
  agentId: string,
  options: { index?: number; ipcDir?: string } = {},
): string {
  const index = options.index ?? 0;
  if (process.platform === "win32") {
    return `\\\\.\\pipe\\haap-${agentId}-agent-assembler-${index}`;
  }
  const dir = options.ipcDir ?? defaultIpcDir();
  return path.join(dir, agentId, `agent-assembler-${index}.sock`);
}

/**
 * High-level HAAP agent client. Connect once, invoke many times, close.
 *
 *     const agent = await HawcxAgent.connect(
 *       "/var/run/haap/research-u1/agent-assembler-0.sock",
 *     );
 *     try {
 *       const response = await agent.invoke({
 *         targetRsUrl: "https://api.example.com/search",
 *         httpMethod: "POST",
 *         headers: { "Content-Type": "application/json" },
 *         tool: "search",
 *         action: ["read"],
 *         body: Buffer.from('{"query": "agents"}'),
 *       });
 *     } finally {
 *       agent.close();
 *     }
 *
 * Not thread-safe; for concurrent use, wrap in a queue or open multiple
 * agents.
 */
export class HawcxAgent {
  private constructor(private client: AssemblerClient | null) {}

  /**
   * Open the agent IPC socket at `endpoint` and complete the version handshake.
   *
   * On Unix, `endpoint` is a filesystem path. On Windows, it's a Named Pipe
   * path (`\\\\.\\pipe\\haap-<agent_id>-agent-assembler-<index>`).
   */
  static async connect(
    endpoint: string,
    options: { timeoutMs?: number } = {},
  ): Promise<HawcxAgent> {
    const client = await AssemblerClient.connect(endpoint, options);
    return new HawcxAgent(client);
  }

  /**
   * Resolve the conventional Assembler-agent endpoint for an agent id, then
   * `connect`.
   */
  static connectByAgentId(
    agentId: string,
    options: { index?: number; ipcDir?: string; timeoutMs?: number } = {},
  ): Promise<HawcxAgent> {
    return HawcxAgent.connect(
      defaultEndpointFor(agentId, options),
      options,
    );
  }

  /**
   * Profile E tool call.
   *
   * Forwards a `ToolCallRequest` to the Assembler and returns the decrypted
   * `ToolCallResponse`. Throws `RequestRejected` if the Assembler rejects.
   *
   * Parameters mirror the fields of `haap_ipc::messages::assembler::
   * ToolCallRequest`. `body` maps to the wire field `plaintext_request_body`.
   */
  async invoke(opts: HawcxAgentInvokeOptions): Promise<ToolCallResponse> {
    if (!this.client) throw new Error("agent already closed");
    const requestId = opts.requestId ?? `req-${randomUUID().replace(/-/g, "").slice(0, 16)}`;
    return this.client.invoke({
      requestId,
      targetRsUrl: opts.targetRsUrl,
      httpMethod: (opts.httpMethod ?? "POST").toUpperCase(),
      headers: opts.headers,
      tool: opts.tool ?? "",
      action: opts.action,
      resource: opts.resource,
      constraints: opts.constraints,
      body: opts.body,
      claimedIntentHash: opts.claimedIntentHash,
      toolArguments: opts.toolArguments,
      contentType: opts.contentType,
      transport: opts.transport,
    });
  }

  /**
   * Profile E first hop: forward a clarification answer to the Assembler.
   */
  async sendClarificationAnswer(args: {
    pendingId: string;
    sessionId: number | bigint;
    answerIndex?: number;
    answerText?: string;
  }): Promise<void> {
    if (!this.client) throw new Error("agent already closed");
    await this.client.sendClarificationAnswer(args);
  }

  /** Close the IPC connection. Idempotent. */
  close(): void {
    if (this.client) {
      this.client.close();
      this.client = null;
    }
  }
}

File: python/src/hawcx_haap/agent.py
"""HawcxAgent — Profile E entry point.

The SDK does not spawn the supervisor (that's an operational concern handled
by the ``hx_agentic_sdk`` release tarball or Docker image); it connects to the
Assembler's already-running agent socket and proxies tool calls.

The Python process never holds session keys (``response_key``, ``K_req``,
``K_resp``). All cryptographic operations happen inside the Assembler process;
the SDK exchanges only plaintext request bodies and decrypted response bodies
over the local IPC socket.
"""

from __future__ import annotations

import os
import sys
import uuid
from collections.abc import Iterable
from pathlib import Path
from typing import Any

from hawcx_haap.errors import HawcxError
from hawcx_haap.ipc import (
    AssemblerClient,
    TokenTransport,
    ToolCallRequest,
    ToolCallResponse,
)


def _default_ipc_dir() -> Path:
    """Match ``crates/haap-supervisor/src/paths.rs`` default base directory."""
    runtime = os.environ.get("XDG_RUNTIME_DIR")
    if runtime:
        return Path(runtime) / "hawcx"
    return Path("/tmp/hawcx")


def default_endpoint_for(
    agent_id: str,
    *,
    index: int = 0,
    ipc_dir: Path | None = None,
) -> str:
    """Compute the conventional Assembler agent-socket path for an agent id.

    - Unix:    ``{ipc_dir}/{agent_id}/agent-assembler-{index}.sock``
    - Windows: ``\\\\.\\pipe\\haap-{agent_id}-agent-assembler-{index}``
    """
    if sys.platform == "win32":
        return rf"\\.\pipe\haap-{agent_id}-agent-assembler-{index}"
    base = ipc_dir or _default_ipc_dir()
    return str(base / agent_id / f"agent-assembler-{index}.sock")


class HawcxAgent:
    """Customer-facing handle for HAAP Profile E tool calls.

    Construct via :meth:`connect` (explicit socket path) or
    :meth:`connect_by_agent_id` (path-by-convention from an agent id). All
    cryptography happens in the Assembler; this class is a thin transport for
    :class:`ToolCallRequest` / :class:`ToolCallResponse`.
    """

    def __init__(self, client: AssemblerClient) -> None:
        self._client = client

    @classmethod
    def connect(
        cls,
        endpoint: str,
        *,
        timeout_secs: float | None = 5.0,
    ) -> HawcxAgent:
        """Open the agent IPC socket at ``endpoint`` and complete the handshake."""
        client = AssemblerClient.connect(endpoint, timeout_secs=timeout_secs)
        return cls(client)

    @classmethod
    def connect_by_agent_id(
        cls,
        agent_id: str,
        *,
        index: int = 0,
        ipc_dir: Path | None = None,
        timeout_secs: float | None = 5.0,
    ) -> HawcxAgent:
        """Resolve the conventional agent-Assembler endpoint, then ``connect``."""
        return cls.connect(
            default_endpoint_for(agent_id, index=index, ipc_dir=ipc_dir),
            timeout_secs=timeout_secs,
        )

    def invoke(
        self,
        *,
        target_rs_url: str,
        http_method: str = "POST",
        headers: dict[str, str] | None = None,
        tool: str = "",
        action: Iterable[str] | None = None,
        resource: str = "*",
        constraints: dict[str, Any] | None = None,
        body: bytes | None = None,
        claimed_intent_hash: str | None = None,
        tool_arguments: Any = None,
        content_type: str | None = None,
        transport: TokenTransport | None = None,
        request_id: str | None = None,
    ) -> ToolCallResponse:
        """Profile E tool call.

        Forwards a :class:`ToolCallRequest` to the Assembler and returns the
        decrypted :class:`ToolCallResponse`. Raises
        :class:`hawcx_haap.errors.RequestRejected` if the Assembler rejects.

        Parameters mirror the fields of ``haap_ipc::messages::assembler::
        ToolCallRequest``. ``body`` maps to the wire field
        ``plaintext_request_body``.
        """
        if self._client is None:
            raise HawcxError("agent already closed")
        req = ToolCallRequest(
            request_id=request_id or f"req-{uuid.uuid4().hex[:16]}",
            target_rs_url=target_rs_url,
            http_method=http_method.upper(),
            headers=dict(headers or {}),
            tool=tool,
            action=list(action or []),
            resource=resource,
            constraints=dict(constraints or {}),
            plaintext_request_body=body,
            claimed_intent_hash=claimed_intent_hash,
            tool_arguments=tool_arguments,
            content_type=content_type,
            transport=transport,
        )
        return self._client.invoke(req)

    def send_clarification_answer(
        self,
        *,
        pending_id: str,
        session_id: int,
        answer_index: int | None = None,
        answer_text: str | None = None,
    ) -> None:
        """Profile E first hop: forward a clarification answer to the Assembler."""
        if self._client is None:
            raise HawcxError("agent already closed")
        self._client.send_clarification_answer(
            pending_id=pending_id,
            session_id=session_id,
            answer_index=answer_index,
            answer_text=answer_text,
        )

    def close(self) -> None:
        if self._client is not None:
            self._client.close()
            self._client = None  # type: ignore[assignment]

    def __enter__(self) -> HawcxAgent:
        return self

    def __exit__(self, *_: Any) -> None:
        self.close()
